sphinx/addnodes.py

```diff
@@ -342,6 +342,51 @@ class pending_xref(nodes.Inline, nodes.Element):
     """
 
 
+class pending_xref_condition(nodes.Inline, nodes.TextElement):
+    """Node for cross-references that are used to choose appropriate
+    content of the reference by conditions on the resolving phase.
+
+    When the :py:class:`pending_xref` node contains one or more
+    **pending_xref_conditions** nodes, the cross-reference resolver
+    should choose the content of the reference using defined conditions
+    in ``condition`` attribute of each pending_xref_conditions node::
+
+        <pending_xref refdomain="py" reftarget="io.StringIO ...>
+            <pending_xref_condition condition="resolved">
+                <literal>
+                    StringIO
+            <pending_xref_condition condition="*">
+                <literal>
+                    io.StringIO
+
+    After the processing of cross-reference resolver, one of the content node
+    under pending_xref_condition node is chosen by its condition and to be
+    removed all of pending_xref_condition nodes::
+
+        # When resolved the cross-reference successfully
+        <reference>
+            <literal>
+                StringIO
+
+        # When resolution is failed
+        <reference>
+            <literal>
+                io.StringIO
+
+    .. note:: This node is only allowed to be placed under pending_xref node.
+              It is not allows to place it under other nodes.  In addition,
+              pending_xref node must contain only pending_xref_condition
+              nodes if it contains one or more pending_xref_condition nodes.
+
+    The pending_xref_condition node should have **condition** attribute.
+    Domains can be store their individual conditions into the attribute to
+    filter contents on resolving phase.  As a reserved condition name,
+    ``condition="*"`` is used for the fallback of resolution failure.
```

Comment:
Now I'm thinking about how intersphinx treats this. I have two idea: 1) adding another reserved condition `"resolved"` as a common hint for extensions, 2) add a filtering method to the domain class.

Reply:
I discussed this with @tk0miya at the Sphinx Hack-a-thon event.

* We can use any name for the condition, but there are reserved and recommended names.
* The 'resolved' is the recommended condition. The reason for the recommendation is that it will be selected if it matches in intersphinx.
* '*' is a reserved condition. It will be selected if no reference is found in the domain or intersphinx.

So, for now, we should choose (1) for this feature.

Reply:
updated.